### PR TITLE
Aufs to overlay2 migration utility

### PR DESCRIPTION
**Pull request by robertgzr**
Migrate image store, containers and daemon configs to overlayfs (overlay2 graphdriver)

### CODE REVIEW

File: cmd/a2o-migrate/a2o/commit.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+	// remove images
+	aufsImageDir := filepath.Join(StorageRoot, "image", "aufs")
+	err = removeDirIfExists(aufsImageDir)
+	if err != nil {
+		return err
+	}
+
+	// remove hostapps
+	hostappsDir := filepath.Join(StorageRoot, "hostapps")
```
Comment (roman-mazur) [inline cmd/a2o-migrate/a2o/commit.go]: This seems to be very balenaOS specific.
Should we consider parameterizing this from outside? For instance, our update scripts in the OS could specify extra directories to remove on commit.

File: cmd/a2o-migrate/a2o/migrate.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+		logrus.Warn("overlay root found, cleaning up...")
+		err := os.Remove(overlayRoot())
+		if err != nil {
+			return fmt.Errorf("Error cleaning up overlay2 root: %v", err)
+		}
+	}
+
+	var (
+		state State
+	)
```
Comment (roman-mazur) [inline cmd/a2o-migrate/a2o/migrate.go]: nit: could be `var state State` (in a single line)

File: cmd/a2o-migrate/a2o/utils.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+func switchAllContainersStorageDriver(newStorageDriver string) error {
+	containerDir := filepath.Join(StorageRoot, "containers")
+	containerIDs, err := osutil.LoadIDs(containerDir)
+	if err != nil {
+		return fmt.Errorf("Error listing containers: %v", err)
+	}
+	logrus.Infof("migrating %v container(s) to overlay2", len(containerIDs))
+	for _, containerID := range containerIDs {
+		err := switchContainerStorageDriver(containerID, "overlay2")
```
Comment (roman-mazur) [inline cmd/a2o-migrate/a2o/utils.go]: Should not the `newStorageDriver` variable be used here?

File: cmd/a2o-migrate/a2o/utils.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		return err
+	}
+	defer f.Close()
+
+	var containerConfig = make(map[string]interface{})
+	err = json.NewDecoder(f).Decode(&containerConfig)
+	if err != nil {
+		return err
+	}
+	containerConfig["Driver"] = "overlay2"
```
Comment (roman-mazur) [inline cmd/a2o-migrate/a2o/utils.go]: Same here.

File: cmd/a2o-migrate/aufsutil/aufsutils.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
+// TODO: do we need to handle .wh..wh.plnk layer hardlinks?
```
Comment (roman-mazur) [inline cmd/a2o-migrate/aufsutil/aufsutils.go]: Do we have an answer to the question? :)

File: cmd/a2o-migrate/cli/main.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+var ( // auto generated on build
+	GitVersion = "undefined"
+	BuildTime  = "undefined"
+)
+
+func main() {
+	a2omigrate.GitVersion = GitVersion
+	a2omigrate.BuildTime = BuildTime
```
Comment (roman-mazur) [inline cmd/a2o-migrate/cli/main.go]: Why do we override them here instead of setting directly those variables with build flags?

File: cmd/a2o-migrate/test/a2o-migrate.sh [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+fi
+
+log "Using $DOCKERDIR as $ENGINE root directory."
+
+# TODO: handle containers as well
+clean_containers
+
+mkdir -p "$DOCKERDIR/overlay2/l"
+log "Migrating layers ..."
+for aufs_layerpath in "$DOCKERDIR"/aufs/diff/*; do
```
Comment (roman-mazur) [inline cmd/a2o-migrate/test/a2o-migrate.sh]: Is this script an analog/alternative to the a2o tool?

File: cmd/a2o-migrate/test/integration.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    # stop aufs daemon
+    $RT stop -t 3 balena
+
+    # start balenaEngine with overlay2
+    $RT run $balena_container_flags $IMAGE --debug --storage-driver=overlay2
+    sleep 1
+    $RT inspect balena &>/dev/null || exit 1
+
+    # check if we still are able to create a container from the a2o-test image
+    $RT exec balena balena-engine run --rm a2o-test ls -lR /tmp
```
Comment (roman-mazur) [inline cmd/a2o-migrate/test/integration.sh]: Shouldn't we check the output here to see it matches the expected value?

File: cmd/a2o-migrate/test/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#FROM balena/armv7hf-supervisor:v9.0.1
+FROM busybox
```
Comment (roman-mazur) [inline cmd/a2o-migrate/test/Dockerfile]: Does it matter what base image is used for the test?

File: vendor/github.com/containerd/containerd/cmd/containerd-shim/main_unix.go [hunk 10]
```diff
@@ -62,7 +62,7 @@ var (
 	containerdBinaryArgv0Flag string
 )
 
-func init() {
+func Main() {
```
Comment (roman-mazur) [inline vendor/github.com/containerd/containerd/cmd/containerd-shim/main_unix.go]: Is this change related?

File: cmd/a2o-migrate/test/a2o-migrate.sh (hunk 7 above)
Comment (robertgzr) [reply cmd/a2o-migrate/test/a2o-migrate.sh]: yes. this was the original bash implementation, I will remove it

File: cmd/a2o-migrate/test/Dockerfile (hunk 9 above)
Comment (robertgzr) [reply cmd/a2o-migrate/test/Dockerfile]: not really I just wanted a smaller image that gets pulled for the test

File: vendor/github.com/containerd/containerd/cmd/containerd-shim/main_unix.go (hunk 10 above)
Comment (robertgzr) [reply vendor/github.com/containerd/containerd/cmd/containerd-shim/main_unix.go]: this was already PR'd into balena-containerd here: https://github.com/balena-os/balena-containerd/pull/5

this commit updates the vendored copy: https://github.com/balena-os/balena-engine/pull/168/commits/9898babd1a7906365003dd940fa784a0c5a77cdb

File: cmd/a2o-migrate/a2o/commit.go (hunk 1 above)
Comment (robertgzr) [reply cmd/a2o-migrate/a2o/commit.go]: good catch :heavy_check_mark:

File: cmd/a2o-migrate/cli/main.go (hunk 6 above)
Comment (robertgzr) [reply cmd/a2o-migrate/cli/main.go]: I guess this is leftover of the restructure... `cmd/a2o-migrate` is built in a way that it can be compiled into github.com/balena-os/balena-engine-cli and it also has the `GitVersion`/`BuildTime` variables [here](https://github.com/balena-os/balena-engine/blob/a2o-migrate/cmd/a2o-migrate/a2o-migrate.go#L13-L15)

I wanted to be able to build it as a standalone tool using go modules though, which is also the reason for those [overrides](https://github.com/balena-os/balena-engine/blob/a2o-migrate/cmd/a2o-migrate/go.mod#L13-L17) and the `vendor/`

Do you have any suggestions what to do with that?

Comment (roman-mazur) [reply cmd/a2o-migrate/cli/main.go]: Cannot we just set the `a2omigrate` package name on the variables we set from the Makefile?
https://github.com/balena-os/balena-engine/pull/168/files#diff-54936c3e2e129070063e5d3725547463R3

File: cmd/a2o-migrate/test/Dockerfile (hunk 9 above)
Comment (roman-mazur) [reply cmd/a2o-migrate/test/Dockerfile]: So should we remove the commented line?

File: cmd/a2o-migrate/test/integration.sh [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+CONTAINERIZED=${CONTAINERIZED:-}
+PROJECT="$(dirname $(readlink -f $0))/.."
+IMAGE=${IMAGE:-balena/balena-engine:beind}
+
+container_name="balena_a2o_integration_test"
+balena_container_flags="--rm --detach --name ${container_name} --privileged -v varlibbalena:/var/lib/balena-engine -v ${PROJECT}:/src:ro -w /src"
+
+set -ex
+
+[ -n "$CONTAINERIZED" ] && {
```
Comment (roman-mazur) [inline cmd/a2o-migrate/test/integration.sh]: Maybe simply move it to a separate file? This part does not use any of the variables above, so it can be cleaner/easier to call a separate script that calls migration commands in the container.

File: contrib/beind/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		xfsprogs \
+		xz \
+# pigz: https://github.com/moby/moby/pull/35697 (faster gzip implementation)
+		pigz \
+	; \
+# only install zfs if it's available for the current architecture
+# https://git.alpinelinux.org/cgit/aports/tree/main/zfs/APKBUILD?h=3.6-stable#n9 ("all !armhf !ppc64le" as of 2017-11-01)
+# "apk info XYZ" exits with a zero exit code but no output when the package exists but not for this arch
+	if zfs="$(apk info --no-cache --quiet zfs)" && [ -n "$zfs" ]; then \
+		apk add --no-cache zfs; \
```
Comment (roman-mazur) [inline contrib/beind/Dockerfile]: Why do we need zfs?

File: cmd/a2o-migrate/cli/main.go (hunk 6 above)
Comment (robertgzr) [reply cmd/a2o-migrate/cli/main.go]: done

File: contrib/beind/Dockerfile (hunk 12 above)
Comment (robertgzr) [reply contrib/beind/Dockerfile]: cleaned up the remaining extra dependencies

**Comment by robertgzr** [comment]
@roman-mazur I have finally squashed this down :)

**Comment by roman-mazur** [comment]
Is it ok travis build fails? Should we fix it or disable it?

**Comment by robertgzr** [comment]
@roman-mazur fixing the travis build here: https://github.com/balena-os/balena-engine/pull/189

**Comment by robertgzr** [comment]
> simulate a failure and rollback

sounds good. will work on that tomorrow

**Comment by robertgzr** [comment]
@roman-mazur 	should we close this for now and reopen when we move the migration logic to the engine too?

**Comment by robertgzr** [comment]
I created https://github.com/balena-os/balena-engine/issues/205 to track the motivation and progress of this, should we decide to postpone the work.

**Comment by robertgzr** [comment]
one solution I see is to merge the core migration code under pkg/ and remove the a2o-migrate binary

**Comment by robertgzr** [comment]
update state here once there is a decision: https://github.com/balena-os/meta-balena/pull/1577

**Comment by robertgzr** [comment]
This should be reworked in to a procedure that runs on engine startup

**Comment by robertgzr** [comment]
updated this with the new approach. moving the a2o code under `pkg/storagemigration` and implementing it to run when the engine starts.

Right now it's guarded by a check if the storage driver is set to `overlay2` and looks for an env var `BALENA_MIGRATE_OVERLAY`. The logic is mostly identical to what we had before...

It will attempt a migration. If that fails run the fail-cleanup routine and if not commit the changes.

Of course this can't check if the migration was actually successful in the sense that we can run anything...

I think it makes more sense to have the integration testing done using a custom meta-balena image, since the previous vagrant-based approach needs a custom OS anyway

File: daemon/daemon.go [hunk 13]
```diff
@@ -933,6 +934,37 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
...
+		logrus.Info("Starting storage migration from aufs to overlay2")
+		start := time.Now()
+		var err error
+		err = storagemigration.Migrate(config.Root)
+		if err != nil {
+			if err == storagemigration.ErrAuFSRootNotExists ||
+				err == storagemigration.ErrOverlayRootExists {
+				// gracefully handle missing aufs root or existing
+				// overlay root - we have nothing to migrate from
+				logrus.Infof("Storage migartion skipped: %s", err)
```
Comment (klutchell) [inline daemon/daemon.go:951]:
```suggestion
				logrus.Infof("Storage migration skipped: %s", err)
```

File: pkg/storagemigration/overlayutils.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+var (
+	// ErrOverlayRootNotExists indicates the overlay2 root directory wasn't found
+	ErrOverlayRootNotExists = errors.New("Overlay2 root doesn't exists")
+	// ErrOverlayRootExists indicates the migration was already run sucessfully
+	ErrOverlayRootExists = errors.New("Overlay2 root exists")
+)
+
+// CheckRootExists checks for the overlay storage root directory
```
Comment (klutchell) [inline pkg/storagemigration/overlayutils.go:23]:
```suggestion
// CheckOverlayRootExists checks for the overlay storage root directory
```

**Comment by robertgzr** [comment]
@klutchell I think I want to merge this one pretty soon. Most of the work here has gone through pretty intense manual testing (on the big platforms anyway) and should we identify any follow-up problems they can always be fixed in subsequent PRs

For the most part merging this would allow me to open the meta-balena part of this as a PR pointing to engine master :)

**Comment by robertgzr** [comment]
@balena-ci rebase

**Comment by klutchell** [comment]
@balena-ci rebase

**Review by klutchell — APPROVED** [review]
LGTM!
